app/api/api_v1/endpoints/data.py

```diff
@@ -17,54 +18,158 @@
     """
-    Upload a .csv file (Using the ploutos .csv provided as how a baseline file should be formatted)
+    Upload a .csv file that contains data about weather.
+
+    List of columns that this api will parse:
+
+    [
+        "date", "time", "parcel_location", "atmospheric_temperature", "atmospheric_temperature_daily_min",
+        "atmospheric_temperature_daily_max", "atmospheric_temperature_daily_average", "atmospheric_relative_humidity",
+        "atmospheric_pressure", "precipitation", "average_wind_speed", "wind_direction", "wind_gust",
+        "leaf_relative_humidity", "leaf_temperature", "leaf_wetness",
+        "soil_temperature_10cm", "soil_temperature_20cm", "soil_temperature_30cm", "soil_temperature_40cm",
+        "soil_temperature_50cm", "soil_temperature_60cm", "solar_irradiance_copernicus"
+    ]
+
+    date and time are mandatory fields.
+
+    date format: %Y-%m-%d
+    time format: %H:%M:%S
+
+    any decimal numbers present in the dataset should be formatted using "." (dot) and not "," (comma),
+    but nevertheless, the api will attempt to swap "," for ".".
+
     """
 
+    dataset_db = crud.dataset.get_by_name(db=db, name=csv_file.filename)
+
+    if dataset_db:
+        raise HTTPException(
+            status_code=400,
+            detail="Error, dataset with same filename already uploaded, please rename your dataset"
+        )
+
     csv_reader = reader(iterdecode(csv_file.file, "utf-8-sig"), delimiter=";")
-    first_column = True
+
+    # Parse the .csv headers
+    headers = next(csv_reader)
+
+    possible_column_names = utils.possible_column_names
+
+    # If there are more columns than what is expected, decline the dataset
+    if len(headers) > len(possible_column_names):
+        raise HTTPException(
+            status_code=400,
+            detail="Error, dataset has more than {} columns, not supported, please conform to required dataset format".format(len(possible_column_names))
+        )
+
+    # if there is no useful information then decline the dataset (see comment #A)
+    if len(headers) < 3:
+        raise HTTPException(
+            status_code=400,
+            detail="Error, won't accept dataset with less than 3 columns, please provide more information in the dataset"
+        )
+
+    if "date" not in headers or "time" not in headers:
+        raise HTTPException(
+            status_code=400,
+            detail="Error, can't upload dataset with no date or time information"
+        )
+
+    # Find usable columns and their place in the file (column sequence)
+    usable_column_names = {}
+
+    col_pos = 0
+    for col in headers:
+        if col in possible_column_names:
+            usable_column_names[col] = col_pos
+        col_pos = col_pos + 1
+
+    #A
+    # Refuse to accept dataset if there is only date+time information.
+    # If but one of these is missing, then we have date+something or time+something which is again, useless.
+    if len(usable_column_names.items()) < 3:
```

Comment:
This condition checks whether there is enough information in this csv file (after it has been "sanitized") to warrant it's upload.

For example, if you have a csv file with the following columns:

date;time;radiation_amount;wind_velocity

after sanitizing, the usable_column_names.items() would hold "date" and "time" since "radiation_amount" and "wind_velocity" would get filtered out by the loop before it.
Since these are the only usable columns left, this dataset wouldn't be able to be used in the service (you can't calculate a probability using date and time with no measurement information)

If for example you had something like this:

date;time;atmospheric_pressure;radiation_amount;wind_velocity

after sanitizing, you would be left with "date", "time" and "atmospheric_pressure", which can be used when calculating risk, because you actually have some useful data via the atmospheric_pressure

I hope this clears it up a bit, if it's a bit much we can organize a call and go through the whole api so that I can better explain why I did some of the checks

Reply:
all clear. thanks!